### PR TITLE
Make HF implementation match original OLMo 2 models for lower precisions

### LINKED ISSUE #38117: OLMo and OLMo 2 models do not match original models for low precisions
### System Info

- `transformers` version: 4.52.0.dev0
- Platform: macOS-14.5-arm64-arm-64bit
- Python version: 3.12.9
- Huggingface_hub version: 0.31.2
- Safetensors version: 0.5.2
- Accelerate version: 1.6.0
- Accelerate config:    not found
- DeepSpeed version: not installed
- PyTorch version (GPU?): 2.6.0 (False)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using distributed or parallel set-up in script?: No

### Who can help?

@ArthurZucker 

### Information

- [ ] The official example scripts
- [x] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

Steps to reproduce behavior:

1. Obtain a copy of an OLMo or OLMo 2 model and its corresponding HF model.
2. Run the original and HF model in a lower precision like float16.

Giving code for a repro is particularly complicated due to running the models in the original codebase, but I intend to fix this bug myself so hopefully it is fine to omit this.

### Expected behavior

Expected behavior is that the output logits match or are close (say, all within 1e-4). In practice, this is not the case for bfloat16 and float16.

```
...
[2025-05-13 16:54:03] INFO     [__main__:246, rank=0] blocks.11.feed_forward_norm|output, model.layers.11.post_feedforward_layernorm|output element diff abs mean: 0.0077401818707585335
[2025-05-13 16:54:03] INFO     [__main__:246, rank=0] lm_head.norm|input, model.norm|input element diff abs mean: 0.015348772518336773
[2025-05-13 16:54:03] INFO     [__main__:246, rank=0] lm_head.norm|output, model.norm|output element diff abs mean: 0.01983717642724514
[2025-05-13 16:54:03] INFO     [__main__:246, rank=0] lm_head.w_out|input, lm_head|input element diff abs mean: 0.01983717642724514
[2025-05-13 16:54:03] INFO     [__main__:234, rank=0] lm_head.w_out|output, lm_head|output shape mismatch: torch.Size([1, 120, 100352]) torch.Size([1, 120, 100278])
[2025-05-13 16:54:03] INFO     [__main__:246, rank=0] lm_head.w_out|output, lm_head|output element diff abs mean: 0.03660527244210243
```

### PR DESCRIPTION
**Pull request by 2015aroras**
# What does this PR do?

Context: We found that the vLLM implementation of OLMo 2 models sometimes NaNs (https://github.com/allenai/olmo-cookbook/issues/60). Investigation revealed that this was caused by vLLM converting float32 models to float16 by default. Follow-up investigation revealed that OLMo 2 models in the original codebase do not match transformers and vLLM models in lower precisions like float16 and bfloat16.

Fix: transformers and original OLMo 2 differ in 2 ways:
1. In RMS norm, weight and hidden states are multiplied before converting back to the input dtype in the original but after in transformers.
2. Rope is applied in full precision in the original OLMo 2 rather than at the default precision. This is also true for OLMo 1.

This PR makes transformers match OLMo 2 in these manners (and OLMo 1 for point 2). This is enough to make the transformers implementation logits match the original OLMo 2's in lower precisions on CPU. This is NOT sufficient for making the transformers implementation logits match the original OLMo 1's logits in lower precisions (reason unknown), but logits still match in float32.

Fixes #38117 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

@ArthurZucker

### CODE REVIEW

File: src/transformers/models/olmo/modeling_olmo.py [hunk 1]
```diff
@@ -302,8 +303,7 @@ def forward(self, x, position_ids):
             emb = torch.cat((freqs, freqs), dim=-1)
             cos = emb.cos() * self.attention_scaling
             sin = emb.sin() * self.attention_scaling
-
-        return cos.to(dtype=x.dtype), sin.to(dtype=x.dtype)
+            return cos, sin
```
Comment (ArthurZucker) [inline src/transformers/models/olmo/modeling_olmo.py:306]: Rope is mega tricky, if you still have issue, I would check the inv_freqs, they are init on device sometimes, which makes a difference

**Review by ArthurZucker — APPROVED** [review]
thanks a lot for upstreaming! happy to merge
